### PR TITLE
Statemachine check

### LINKED ISSUE #436: Dependency analysis for Phase 1 state machines
- [x] `fpp-depend` and unit tests
- [x] `fpp-locate-defs` and unit tests
- [x] `fpp-locate-uses` and unit tests

### LINKED ISSUE #437: Semantics for Phase 1 state machines
- [x] Update semantics
- [x] Update `fpp-check` and unit tests

### PR DESCRIPTION
**Pull request by garthwatney**
Added dependency analysis and semantic checking to state machines Phase 1.

Closes #436. Closes #437.

### CODE REVIEW

**Comment by bocchino** [comment]
Per discussion, we need to check the following rules:
1. A component containing a state machine instance must be active or queued.
2. An active component containing a state machine instance and no other async element is valid.

**Comment by bocchino** [comment]
We also need to check that each state machine instance specifier in a component definition has a distinct name. I added this rule to the spec for component definitions.

**Review by bocchino — COMMENTED** [review]
Looks great! Here are some comments:

1. We need to update locate-defs in addition to locate-uses.
2. We need unit tests for state machine defs and state machine instance specs that are similar to the existing unit tests for type defs and command/event specs.

File: compiler/lib/src/main/scala/analysis/UsedSymbols.scala [hunk 1]
```diff
@@ -61,6 +67,8 @@ object UsedSymbols extends UseAnalyzer {
         case Symbol.EnumConstant(node) => defEnumConstantAnnotatedNode(a1, node)
         case Symbol.Module(node) => defModuleAnnotatedNode(a1, node)
         case Symbol.Port(node) => defPortAnnotatedNode(a1, node)
+        case Symbol.StateMachine(node) => defStateMachineAnnotatedNode(a1, node)
+        case Symbol.StateMachineInstance(node) => specStateMachineInstanceAnnotatedNode(a1, node)
```
Comment (bocchino) [inline compiler/lib/src/main/scala/analysis/UsedSymbols.scala:71]: There should not be a symbol for a state machine instance. Only definitions get symbols.

File: compiler/lib/src/main/scala/analysis/CheckSemantics/EnterSymbols.scala [hunk 2]
```diff
@@ -217,6 +217,32 @@ object EnterSymbols
       yield updateMap(a, symbol).copy(nestedScope = nestedScope)
   }
```
Comment (bocchino) [inline compiler/lib/src/main/scala/analysis/CheckSemantics/EnterSymbols.scala:219]: We shouldn't enter a symbol here. We only enter symbols for definitions.

File: compiler/lib/src/main/scala/analysis/UsedSymbols.scala (hunk 1 above)
Comment (garthwatney) [reply compiler/lib/src/main/scala/analysis/UsedSymbols.scala]: Fixed

File: compiler/lib/src/main/scala/analysis/Semantics/StateMachineInstance.scala [hunk 3]
```diff
@@ -0,0 +1,28 @@
+package fpp.compiler.analysis
+
+import fpp.compiler.ast._
+import fpp.compiler.util._
+
+/** An FPP data product containter */
```
Comment (bocchino) [inline compiler/lib/src/main/scala/analysis/Semantics/StateMachineInstance.scala:6]: Looks like a stray comment: "data product container" should be "state machine instance"

File: compiler/lib/src/main/scala/analysis/Semantics/StateMachineInstance.scala [hunk 4]
```diff
@@ -0,0 +1,28 @@
...
+  aNode: Ast.Annotated[AstNode[Ast.SpecStateMachineInstance]]
+) {
+
+  /** Gets the location of the state machine */
+  def getLoc: Location = Locations.get(aNode._2.id)
+
+  /** Gets the unqualified name of the state machine instance */
+  def getUnqualifiedName = aNode._2.data.name
+
+
```
Comment (bocchino) [inline compiler/lib/src/main/scala/analysis/Semantics/StateMachineInstance.scala:17]: Let's close up the extra blank line

File: compiler/lib/src/main/scala/analysis/Semantics/StateMachineInstance.scala [hunk 5]
```diff
@@ -0,0 +1,28 @@
...
+import fpp.compiler.ast._
+import fpp.compiler.util._
+
+/** An FPP data product containter */
+final case class StateMachineInstance(
+  aNode: Ast.Annotated[AstNode[Ast.SpecStateMachineInstance]]
+) {
+
+  /** Gets the location of the state machine */
```
Comment (bocchino) [inline compiler/lib/src/main/scala/analysis/Semantics/StateMachineInstance.scala:11]: "state machine" --> "state machine instance"

File: compiler/lib/src/main/scala/codegen/LocateDefsFppWriter.scala [hunk 6]
```diff
@@ -50,6 +50,16 @@ object LocateDefsFppWriter extends AstVisitor with LineUtils {
       data.members.flatMap(matchComponentMember(s1, _))
   }
 
+  override def defStateMachineAnnotatedNode(
+    s: State,
+    aNode: Ast.Annotated[AstNode[Ast.DefStateMachine]]
+  ) = {
+    val (_, node, _) = aNode
+    val data = node.data
+    val s1 = s.copy(scopeNameList = data.name :: s.scopeNameList)
```
Comment (bocchino) [inline compiler/lib/src/main/scala/codegen/LocateDefsFppWriter.scala:59]: The line `val s1 = ...` doesn't do anything. Let's delete it. We don't need to visit the members of a state machine because (1) there aren't any in Phase 1 and (2) even in Phase 2 they won't contain any definitions that have locations.

File: compiler/lib/src/main/scala/syntax/Parser.scala [hunk 7]
```diff
@@ -471,6 +471,7 @@ object Parser extends Parsers {
       constant ^^ { case _ => Ast.SpecLoc.Constant } |
       instance ^^ { case _ => Ast.SpecLoc.ComponentInstance } |
       port ^^ { case _ => Ast.SpecLoc.Port } |
+      state ~ machine ^^ { case _ => Ast.SpecLoc.StateMachine } |
```
Comment (bocchino) [inline compiler/lib/src/main/scala/syntax/Parser.scala:474]: Delete this. It's not a correct lexical element.

File: compiler/lib/src/main/scala/util/Error.scala [hunk 8]
```diff
@@ -48,6 +48,9 @@ sealed trait Error {
       case SemanticError.DuplicateDictionaryName(kind, name, loc, prevLoc) =>
         Error.print (Some(loc)) (s"duplicate ${kind} name ${name}")
         printPrevLoc(prevLoc)
+      case SemanticError.DuplicateStateMachineInstance(name, loc, prevLoc) =>
+        Error.print (Some(loc)) (s"duplicate name ${name}")
```
Comment (bocchino) [inline compiler/lib/src/main/scala/util/Error.scala:52]: Let's say "duplicate state machine instance name" in the error message. Also, let's keep the case alphabetized.

File: compiler/lib/src/main/scala/util/Error.scala [hunk 9]
```diff
@@ -206,6 +209,8 @@ sealed trait Error {
         System.err.println(loc2)
       case SemanticError.PassiveAsync(loc) =>
         Error.print (Some(loc)) ("passive component may not have async input")
+      case SemanticError.PassiveStateMachine(loc) =>
```
Comment (bocchino) [inline compiler/lib/src/main/scala/util/Error.scala:212]: I think the error should be PassiveStateMachineInstance. I think a passive component can have a state machine definition, right?

File: compiler/lib/src/main/scala/util/Error.scala [hunk 10]
```diff
@@ -206,6 +209,8 @@ sealed trait Error {
...
+        Error.print (Some(loc)) ("passive component may not have a state machine")
```
Comment (bocchino) [inline compiler/lib/src/main/scala/util/Error.scala:213]: I think the error should say "passive component may not have a state machine instance."

File: compiler/lib/src/main/scala/util/Version.scala [hunk 11]
```diff
@@ -3,6 +3,6 @@ package fpp.compiler.util
 /** The compiler version */
 object Version {
 
-  val v = "[unknown version]"
+  val v = "v2.1.0a10-139-g8e0147ac"
```
Comment (bocchino) [inline compiler/lib/src/main/scala/util/Version.scala:6]: Revert this change.

File: compiler/lib/src/main/scala/analysis/CheckSemantics/CheckComponentDefs.scala [hunk 12]
```diff
@@ -127,6 +127,18 @@ object CheckComponentDefs
...
+    a: Analysis,
+    aNode: Ast.Annotated[AstNode[Ast.SpecStateMachineInstance]]
+  ) = {
+    val stateMachineInstance = StateMachineInstance.fromSpecStateMachine(a, aNode)
+    for {
+      component <- a.component.get.addStateMachineInstance(stateMachineInstance)
+    }
+    yield a.copy(component = Some(component))
+  }
+
```
Comment (bocchino) [inline compiler/lib/src/main/scala/analysis/CheckSemantics/CheckComponentDefs.scala:140]: Close up blank line

File: compiler/tools/fpp-check/test/component/passive_statemachine.fpp [hunk 13]
```diff
@@ -0,0 +1,5 @@
+passive component C {
+    state machine LedSm
+    state machine instance led1: LedSm
+
```
Comment (bocchino) [inline compiler/tools/fpp-check/test/component/passive_statemachine.fpp:4]: Extra blank line

File: compiler/tools/fpp-check/test/component/tests.sh [hunk 14]
```diff
@@ -12,6 +12,7 @@ enum_undef_constant
 enum_undef_type
 queued_no_async_input
 ok
+passive_statemachine
```
Comment (bocchino) [inline compiler/tools/fpp-check/test/component/tests.sh:15]: I see active_statemachine.fpp but no test for it. Also let's be consistent: statemachine --> state_machine.

File: compiler/tools/fpp-check/test/redef/tests.sh [hunk 15]
```diff
@@ -17,6 +17,7 @@ module_component_instance
 module_constant
 module_enum
 module_port
+module_statemachine
```
Comment (bocchino) [inline compiler/tools/fpp-check/test/redef/tests.sh:20]: I see component_statemachine.fpp but no test or ref for it. Let's be consistent: statemachine --> state_machine.

File: compiler/tools/fpp-depend/test/def_state_machine.fpp [hunk 16]
```diff
@@ -0,0 +1,26 @@
...
+    state machine instance foo1: Foo
+
+    state machine instance foo2: Foo
+
+    state machine instance bar: StateMachines.Bar
+
+  }
+
+}
+
```
Comment (bocchino) [inline compiler/tools/fpp-depend/test/def_state_machine.fpp:25]: Extra blank lines

File: compiler/tools/fpp-depend/test/def_state_machine.fpp [hunk 17]
```diff
@@ -0,0 +1,26 @@
...
+  port Port1(
+          a: U32
+          b: F64
+          )
+
+  active component C {
+
+    async input port dataIn: Port1
+
+    state machine Foo
```
Comment (bocchino) [inline compiler/tools/fpp-depend/test/def_state_machine.fpp:14]: For stylistic consistency, I would prefer to avoid using Foo and Bar. Can we use letters like S and T? That's what we do in most of the other test cases.

File: compiler/tools/fpp-locate-defs/test/defs/defs-2.fpp [hunk 18]
```diff
@@ -12,12 +12,16 @@ module M {
...
 
+  state machine LedSm
+
   passive component C {
     type T
     array A = [3] U32
     constant a = 0
     enum E { X, Y, Z }
     struct S { x: U32 }
+    state machine LedSm
```
Comment (bocchino) [inline compiler/tools/fpp-locate-defs/test/defs/defs-2.fpp:23]: For stylistic consistency can we just say SM? The Led part is not necessary.

File: compiler/tools/fpp-check/test/component/ok.fpp [hunk 19]
```diff
@@ -39,4 +39,21 @@ active component C {
 
   telemetry port tlmOut
 
+  state machine LedSm
```
Comment (bocchino) [inline compiler/tools/fpp-check/test/component/ok.fpp:42]: For stylistic consistency let's use SM instead of LedSm.

**Review by bocchino — CHANGES_REQUESTED** [review]
Looks good so far! Please see the comments.

There are missing tests and semantic checks.

* We should add a subdirectory `state_machine_instance` to`tools/fpp-check/test`. This should have tests similar to the tests in `port_instance`. It should check all the semantic rules for state machine instances. Some of these tests are scattered elsewhere, but they should be moved here. We need to add a test for multiple state machine instances in the same component with the same name.

* We need to check for and test the error case where the wrong kind of symbol is used as a state machine instance. For example, fpp-check currently accepts this:
  ```
  [I] bocchino@MT-504857 ~/J/T/f/g/f/compiler (statemachine-check)> 
  bin/fpp-check
  module M {}
  active component C {
    state machine instance m: M
  }
  ```
  This behavior is incorrect. The behavior should be similar to what happens in 
  other cases, e.g.,
  ```
  [I] bocchino@MT-504857 ~/J/T/f/g/f/compiler (statemachine-check) [1]> 
  bin/fpp-check
  module M {}
  active component C {
    async input port p: M
  }
  fpp-check
  stdin:3.23
    async input port p: M
                        ^
  error: invalid symbol M: not a port symbol
  symbol is defined here:
  stdin:1.1
  module M {}
  ^
  ```
  The corresponding tests should go in `tools/fpp-check/invalid_symbols`.

File: compiler/lib/src/main/scala/analysis/Semantics/StateMachineInstance.scala (hunk 3 above)
Comment (garthwatney) [reply compiler/lib/src/main/scala/analysis/Semantics/StateMachineInstance.scala:6]: Fixed

File: compiler/lib/src/main/scala/analysis/Semantics/StateMachineInstance.scala (hunk 4 above)
Comment (garthwatney) [reply compiler/lib/src/main/scala/analysis/Semantics/StateMachineInstance.scala:17]: Fixed

File: compiler/lib/src/main/scala/analysis/Semantics/StateMachineInstance.scala (hunk 5 above)
Comment (garthwatney) [reply compiler/lib/src/main/scala/analysis/Semantics/StateMachineInstance.scala:11]: Fixed

File: compiler/lib/src/main/scala/codegen/LocateDefsFppWriter.scala (hunk 6 above)
Comment (garthwatney) [reply compiler/lib/src/main/scala/codegen/LocateDefsFppWriter.scala:59]: Fixed

File: compiler/lib/src/main/scala/syntax/Parser.scala (hunk 7 above)
Comment (garthwatney) [reply compiler/lib/src/main/scala/syntax/Parser.scala:474]: Fixed

File: compiler/lib/src/main/scala/util/Error.scala (hunk 8 above)
Comment (garthwatney) [reply compiler/lib/src/main/scala/util/Error.scala:52]: Fixed

File: compiler/lib/src/main/scala/util/Error.scala (hunk 9 above)
Comment (garthwatney) [reply compiler/lib/src/main/scala/util/Error.scala:212]: I think we should disallow state machine definitions in a passive component because it's confusing as we don't allow state machine instances in a passive component.

File: compiler/lib/src/main/scala/util/Error.scala (hunk 10 above)
Comment (garthwatney) [reply compiler/lib/src/main/scala/util/Error.scala:213]: Fixed

File: compiler/lib/src/main/scala/util/Version.scala (hunk 11 above)
Comment (garthwatney) [reply compiler/lib/src/main/scala/util/Version.scala:6]: Fixed

File: compiler/lib/src/main/scala/analysis/CheckSemantics/CheckComponentDefs.scala (hunk 12 above)
Comment (garthwatney) [reply compiler/lib/src/main/scala/analysis/CheckSemantics/CheckComponentDefs.scala:140]: Fixed

File: compiler/tools/fpp-check/test/component/passive_statemachine.fpp (hunk 13 above)
Comment (garthwatney) [reply compiler/tools/fpp-check/test/component/passive_statemachine.fpp:4]: Fixed

File: compiler/tools/fpp-check/test/component/tests.sh (hunk 14 above)
Comment (garthwatney) [reply compiler/tools/fpp-check/test/component/tests.sh:15]: Renamed to active_state_machine.fpp and passive_state_machine.fpp and added to tests.sh

File: compiler/tools/fpp-check/test/redef/tests.sh (hunk 15 above)
Comment (garthwatney) [reply compiler/tools/fpp-check/test/redef/tests.sh:20]: Fixed

File: compiler/tools/fpp-depend/test/def_state_machine.fpp (hunk 16 above)
Comment (garthwatney) [reply compiler/tools/fpp-depend/test/def_state_machine.fpp:25]: Fixed

File: compiler/tools/fpp-depend/test/def_state_machine.fpp (hunk 17 above)
Comment (garthwatney) [reply compiler/tools/fpp-depend/test/def_state_machine.fpp:14]: I will use S for state machine definitions and s1 and s2 for state machine instances

File: compiler/tools/fpp-locate-defs/test/defs/defs-2.fpp (hunk 18 above)
Comment (garthwatney) [reply compiler/tools/fpp-locate-defs/test/defs/defs-2.fpp:23]: Fixed as specified in previous comment

File: compiler/tools/fpp-check/test/component/ok.fpp (hunk 19 above)
Comment (garthwatney) [reply compiler/tools/fpp-check/test/component/ok.fpp:42]: Fixed as specified in previous comment

File: compiler/lib/src/main/scala/util/Error.scala (hunk 9 above)
Comment (bocchino) [reply compiler/lib/src/main/scala/util/Error.scala:212]: I agree. You can update the spec, update the checking, and add tests.

File: compiler/tools/fpp-depend/test/def_state_machine.fpp (hunk 17 above)
Comment (bocchino) [reply compiler/tools/fpp-depend/test/def_state_machine.fpp:14]: Sounds good!

File: compiler/lib/src/main/scala/syntax/Parser.scala (hunk 7 above)
Comment (garthwatney) [reply compiler/lib/src/main/scala/syntax/Parser.scala:474]: Actually this is needed for the fpp-depend tool.  I'm putting it back in.

Comment (bocchino) [reply compiler/lib/src/main/scala/syntax/Parser.scala:474]: I agree, it's needed. I got confused between the lexer and the parser. Ignore my comment.

**Comment by garthwatney** [comment]
I created a state_machine_instance directory for state machine tests
I added checks for incorrect symbols in state machine instances.

**Review by bocchino — APPROVED** [review]
Looks good! I added some tests.
